File: backend/app/schemas/governance.py
"""Pydantic request and response schemas for Phase 33: AI Governance & Observability."""

from typing import Any, Dict, List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


# =============================================================================
# 1. Traces & Spans
# =============================================================================

class TraceEventResponse(BaseModel):
    id: str
    span_type: str
    name: str
    input_summary: Dict[str, Any] = Field(default_factory=dict)
    output_summary: Dict[str, Any] = Field(default_factory=dict)
    tokens_consumed: int
    duration_ms: float
    status: str
    error_message: Optional[str] = None
    created_at: datetime


class TraceDetailResponse(BaseModel):
    id: str
    tenant_id: str
    workflow_id: str
    project_id: Optional[str] = None
    lead_id: Optional[str] = None
    agent_id: str
    agent_version: str
    model_id: str
    model_version: str
    prompt_version: str
    total_tokens: int
    estimated_cost: float
    total_duration_ms: float
    status: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    events: List[TraceEventResponse] = Field(default_factory=list)


class TraceCreateRequest(BaseModel):
    workflow_id: str
    project_id: Optional[str] = None
    lead_id: Optional[str] = None
    agent_id: str
    agent_version: str = "v1.0"
    model_id: str = "gpt-4o-mini"
    model_version: str = "v1.0"
    prompt_version: str = "v1.0"


# =============================================================================
# 2. Prompts & Versions
# =============================================================================

class PromptVersionResponse(BaseModel):
    id: str
    version: str
    content: str
    content_hash: str
    status: str
    approved_by: Optional[str] = None
    approved_at: Optional[datetime] = None
    created_at: datetime


class PromptItemResponse(BaseModel):
    id: str
    tenant_id: str
    prompt_key: str
    name: str
    agent_target: str
    purpose: str
    current_version: str
    created_at: datetime
    versions: List[PromptVersionResponse] = Field(default_factory=list)


class PromptCreateRequest(BaseModel):
    prompt_key: str
    name: str
    agent_target: str
    purpose: str
    content: str
    version: str = "v1.0"


# =============================================================================
# 3. Evaluations
# =============================================================================

class EvaluationCaseResponse(BaseModel):
    id: str
    title: str
    difficulty: str
    created_at: datetime


class EvaluationDatasetResponse(BaseModel):
    id: str
    tenant_id: str
    dataset_key: str
    name: str
    description: str
    task_type: str
    is_golden: bool
    version: str
    created_at: datetime
    cases: List[EvaluationCaseResponse] = Field(default_factory=list)


class EvaluationRunRequest(BaseModel):
    dataset_id: str
    agent_key: str
    agent_version: str = "v1.0"
    prompt_version: str = "v1.0"
    model_version: str = "v1.0"
    baseline_score: float = 90.0


class EvaluationRunResponse(BaseModel):
    id: str
    tenant_id: str
    dataset_id: str
    agent_key: str
    agent_version: str
    prompt_version: str
    model_version: str
    evaluation_type: str
    overall_score: float
    passed_cases_count: int
    failed_cases_count: int
    regression_detected: bool
    regression_details: Dict[str, Any]
    completed_at: datetime


# =============================================================================
# 4. Human Reviews, Incidents & Kill Switch
# =============================================================================

class HumanEvaluationRequest(BaseModel):
    trace_id: str
    reviewer: str = "user"
    correctness_score: int = Field(default=5, ge=1, le=5)
    completeness_score: int = Field(default=5, ge=1, le=5)
    evidence_score: int = Field(default=5, ge=1, le=5)
    safety_score: int = Field(default=5, ge=1, le=5)
    usefulness_score: int = Field(default=5, ge=1, le=5)
    notes: Optional[str] = None


class AIIncidentResponse(BaseModel):
    id: str
    tenant_id: str
    incident_number: str
    title: str
    severity: str
    status: str
    affected_agent: str
    description: str
    containment_action: Optional[str] = None
    root_cause: Optional[str] = None
    resolved_by: Optional[str] = None
    resolved_at: Optional[datetime] = None
    created_at: datetime


class AIIncidentCreateRequest(BaseModel):
    title: str
    severity: str = "HIGH"
    affected_agent: str
    description: str
    containment_action: Optional[str] = None


class KillSwitchRequest(BaseModel):
    level: str = Field(default="GLOBAL_AI_OFF", description="GLOBAL_AI_OFF, AGENT_OFF, MODEL_OFF, TOOL_OFF")
    target_key: str = Field(default="GLOBAL")
    is_active: bool = True
    activated_by: str = "admin"
    reason: str


class AIHealthSnapshotResponse(BaseModel):
    id: str
    tenant_id: str
    agent_key: str
    status: str
    success_rate_pct: float
    avg_latency_ms: float
    human_acceptance_pct: float
    daily_cost_consumed: float
    captured_at: datetime

File: frontend/lib/api/governance.ts
import { api } from './client';

export interface GovernanceExecutiveSummary {
  governance_health: string;
  composite_compliance_score: number;
  critical_blockers: number;
  governance_debt_score: number;
  active_exceptions: number;
  advisory_recommendation: string;
  generated_at: string;
}

export interface GovernancePostureData {
  composite_compliance_score: number;
  overall_health: string;
  total_requirements: number;
  implemented_requirements: number;
  total_controls: number;
  healthy_controls: number;
  failing_controls: number;
  open_findings_count: number;
  critical_findings_count: number;
  active_exceptions_count: number;
  stale_evidence_count: number;
  domain_scores: Record<string, number>;
  technical_debt_score: number;
  captured_at: string;
}

export interface GovernanceFrameworkItem {
  framework_code: string;
  name: string;
  version: string;
  category: string;
  jurisdiction: string;
  description: string;
  status: string;
  owner: string;
}

export interface GovernanceRequirementItem {
  requirement_code: string;
  framework_code: string;
  title: string;
  description: string;
  category: string;
  priority: string;
  status: string;
}

export interface GovernanceControlItem {
  control_code: string;
  name: string;
  objective: string;
  domain: string;
  control_type: string;
  frequency: string;
  automation_level: string;
  owner_id: string;
  operator_id: string;
  health_status: string;
  implementations: string[];
}

export interface GovernanceEvidenceItem {
  evidence_id: string;
  evidence_type: string;
  source_subsystem: string;
  source_record_id: string;
  sha256_hash: string;
  provenance_uri: string;
  summary: string;
  collected_at: string;
  freshness_status: string;
}

export interface GovernanceRiskItem {
  risk_code: string;
  title: string;
  category: string;
  likelihood: number;
  impact: number;
  inherent_risk_score: number;
  residual_risk_score: number;
  risk_level: string;
  treatment: string;
  owner_id: string;
  status: string;
  mitigating_controls: string[];
}

export interface GovernanceExceptionItem {
  exception_code: string;
  title: string;
  control_code: string;
  reason: string;
  compensating_controls: string[];
  owner_id: string;
  requester_id: string;
  approver_id?: string;
  status: string;
  expiration_date: string;
}

export interface GovernanceFindingItem {
  finding_code: string;
  control_code: string;
  title: string;
  description: string;
  severity: string;
  root_cause?: string;
  owner_id: string;
  status: string;
  due_date: string;
}

export interface PrivacyRequestItem {
  request_code: string;
  subject_id: string;
  request_type: string;
  status: string;
  requested_at: string;
  due_date: string;
  assigned_to: string;
}

export interface VendorProfileItem {
  vendor_code: string;
  name: string;
  service_provided: string;
  criticality: string;
  data_access_level: string;
  security_risk: string;
  privacy_risk: string;
  owner_id: string;
  status: string;
  dependent_services: string[];
}

export interface CopilotAnswer {
  query: string;
  answer: string;
  evidence_citations: string[];
  disclaimer: string;
  uncertainty_score: number;
  timestamp: string;
}

export interface TraceEvent {
  id: string;
  name: string;
  span_type: string;
  status: string;
  timestamp?: string;
  created_at?: string;
  input_summary?: any;
  output_summary?: any;
  duration_ms?: number;
  tokens_consumed?: number;
  metadata?: Record<string, any>;
}

export interface AITrace {
  id: string;
  trace_id?: string;
  tenant_id?: string;
  started_at?: string;
  agent_id: string;
  task_name?: string;
  status: string;
  duration_ms?: number;
  tokens_used?: number;
  events?: TraceEvent[];
  created_at?: string;
  total_duration_ms?: number;
  agent_version?: string;
  workflow_id?: string;
  model_id?: string;
  model_version?: string;
  total_tokens?: number;
  estimated_cost?: number;
  prompt_version?: string;
}

export interface PromptItem {
  id: string;
  prompt_code?: string;
  name: string;
  version?: string;
  template?: string;
  status?: string;
  author?: string;
  created_at: string;
  tenant_id?: string;
  current_version?: string;
  purpose?: string;
  agent_target?: string;
  prompt_key?: string;
  versions?: any[];
}

export interface EvaluationDataset {
  id: string;
  dataset_code?: string;
  name: string;
  sample_count?: number;
  domain?: string;
  created_at: string;
  tenant_id?: string;
  dataset_key?: string;
  description?: string;
  cases?: any[];
  task_type?: string;
  is_golden?: boolean;
  version?: string;
}

export interface EvaluationRun {
  id: string;
  run_code?: string;
  dataset_id?: string;
  model_name?: string;
  pass_rate_pct?: number;
  latency_p99_ms?: number;
  status?: string;
  completed_at?: string;
  tenant_id?: string;
  agent_key?: string;
  agent_version?: string;
  prompt_version?: string;
  model_version?: string;
  overall_score?: number;
  passed_cases_count?: number;
  failed_cases_count?: number;
  regression_detected?: boolean;
  regression_details?: string | Record<string, any>;
  evaluation_type?: string;
}

export interface AIIncident {
  id: string;
  incident_code: string;
  title: string;
  severity: string;
  status: string;
  affected_agent_id: string;
  detected_at: string;
}

export interface KillSwitchEvent {
  id: string;
  switch_code: string;
  target_agent_id: string;
  reason: string;
  triggered_by: string;
  is_active: boolean;
  triggered_at: string;
  level?: string;
}

export const governanceApi = {
  getOverview: () => api<GovernanceExecutiveSummary>('/governance/overview'),
  getPosture: () => api<GovernancePostureData>('/governance/posture'),

  listFrameworks: () => api<GovernanceFrameworkItem[]>('/governance/frameworks'),
  listRequirements: (frameworkCode: string = 'SOC2_TYPE_II') =>
    api<GovernanceRequirementItem[]>(`/governance/requirements?framework_code=${frameworkCode}`),

  listControls: (domain?: string) =>
    api<GovernanceControlItem[]>(domain ? `/governance/controls?domain=${domain}` : '/governance/controls'),

  testControl: (controlCode: string, payload: { test_type: string; result: string; details: string; executed_by: string }) =>
    api<any>(`/governance/controls/${controlCode}/test`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),

  listRisks: (category?: string) =>
    api<GovernanceRiskItem[]>(category ? `/governance/risks?category=${category}` : '/governance/risks'),

  acceptRisk: (riskCode: string, approverId: string, justification: string) =>
    api<GovernanceRiskItem>(`/governance/risks/${riskCode}/accept`, {
      method: 'POST',
      body: JSON.stringify({ approver_id: approverId, justification }),
    }),

  listExceptions: () => api<GovernanceExceptionItem[]>('/governance/exceptions'),
  requestException: (payload: any) =>
    api<GovernanceExceptionItem>('/governance/exceptions', {
      method: 'POST',
      body: JSON.stringify(payload),
    }),

  approveException: (exceptionCode: string, approverId: string) =>
    api<GovernanceExceptionItem>(`/governance/exceptions/${exceptionCode}/approve`, {
      method: 'POST',
      body: JSON.stringify({ approver_id: approverId }),
    }),

  listFindings: () => api<GovernanceFindingItem[]>('/governance/findings'),
  createRemediationPlan: (findingCode: string, planTitle: string, actions: any[], ownerId: string) =>
    api<any>(`/governance/findings/${findingCode}/remediation`, {
      method: 'POST',
      body: JSON.stringify({ plan_title: planTitle, actions, owner_id: ownerId }),
    }),

  verifyRemediation: (findingCode: string, verifierId: string, evidenceId: string, retestPassed: boolean) =>
    api<any>(`/governance/findings/${findingCode}/verify`, {
      method: 'POST',
      body: JSON.stringify({
        verifier_id: verifierId,
        verification_evidence_id: evidenceId,
        retest_passed: retestPassed,
      }),
    }),

  listPrivacyRequests: () => api<PrivacyRequestItem[]>('/governance/privacy/requests'),
  listVendors: () => api<VendorProfileItem[]>('/governance/vendors'),

  askCopilot: (query: string) =>
    api<CopilotAnswer>(`/governance/copilot/query?query=${encodeURIComponent(query)}`, {
      method: 'POST',
    }),

  listTraces: async (): Promise<AITrace[]> => [],
  listPrompts: async (): Promise<PromptItem[]> => [],
  listEvaluationDatasets: async (): Promise<EvaluationDataset[]> => [],
  listEvaluationRuns: async (): Promise<EvaluationRun[]> => [],
  listIncidents: async (): Promise<AIIncident[]> => [],
  getActiveKillSwitches: async (): Promise<KillSwitchEvent[]> => [],
  runEvaluationBenchmark: async (payload: any): Promise<any> => ({}),
  triggerKillSwitch: async (payload: any): Promise<any> => ({}),
  registerPrompt: async (payload: any): Promise<any> => ({}),
};
